Implement FocusOptions.focusVisible

lukewarlow opened the pull request:
#### 20aaacc363ae8d250a9b0c203c4053b2818abfe4
<pre>
Implement FocusOptions.focusVisible
<a href="https://bugs.webkit.org/show_bug.cgi?id=242456">https://bugs.webkit.org/show_bug.cgi?id=242456</a>

Reviewed by Manuel Rego Casasnovas.

Implement the focusVisible item in FocusOptions.

When this value is provided the focus visibility is enabled or disabled, if the value isn&apos;t provided
we fallback to an implementation defined value.

An assertion is removed as the case that&apos;s being protected against is now valid.

* LayoutTests/imported/w3c/web-platform-tests/html/interaction/focus/processing-model/focusVisible-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setFocus):
(WebCore::Element::setHasFocusVisible):
(WebCore::Element::focus):
* Source/WebCore/dom/FocusOptions.h:
* Source/WebCore/dom/FocusOptions.idl:

Canonical link: <a href="https://commits.webkit.org/287959@main">https://commits.webkit.org/287959@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4db5967b28aba90466ac9660f2f4bd2b1410d404

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/392 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85398 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31854 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82981 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/410 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8192 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63151 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20923 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83939 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73605 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43454 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/639 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27765 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30311 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71685 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28303 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86830 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5717 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71453 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8274 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69441 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70693 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17711 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14719 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13646 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8059 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13580 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7898 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11417 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9704 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4114,7 +4113,9 @@ void Element::focus(const FocusOptions& options)
         FocusOptions optionsWithVisibility = options;
-        if (options.trigger == FocusTrigger::Bindings && document->wasLastFocusByClick())
+        if (options.focusVisible.has_value())
+            optionsWithVisibility.visibility = options.focusVisible.value() ? FocusVisibility::Visible : FocusVisibility::ForceInvisible;
```
```suggestion
        if (options.focusVisible)
            optionsWithVisibility.visibility = *options.focusVisible ? FocusVisibility::Visible : FocusVisibility::ForceInvisible;
```

mrego reviewed (approved):
r+

lukewarlow committed:
Implement FocusOptions.focusVisible
https://bugs.webkit.org/show_bug.cgi?id=242456

Reviewed by Manuel Rego Casasnovas.

Implement the focusVisible item in FocusOptions.

When this value is provided the focus visibility is enabled or disabled, if the value isn't provided
we fallback to an implementation defined value.

An assertion is removed as the case that's being protected against is now valid.

* LayoutTests/imported/w3c/web-platform-tests/html/interaction/focus/processing-model/focusVisible-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setFocus):
(WebCore::Element::setHasFocusVisible):
(WebCore::Element::focus):
* Source/WebCore/dom/FocusOptions.h:
* Source/WebCore/dom/FocusOptions.idl:

Canonical link: https://commits.webkit.org/287959@main
